tensorflow/core/kernels/non_max_suppression_op.cu.cc

```diff
@@ -0,0 +1,487 @@
+      return;                                                          \
+    }                                                                  \
+  } while (0)
+
+struct __align__(16) Box {
+  float x1, y1, x2, y2;
+};
+
+namespace tensorflow {
+typedef Eigen::GpuDevice GPUDevice;
+typedef Eigen::ThreadPoolDevice CPUDevice;
+
+// This is the width of the bitmask for masking boxes for each thread.
+// This needs to be a multiple of 2(a POD width usually) so that division and
+// modulo can be implemented as bit operations during host selection.
+constexpr int kNmsBoxesPerThread = 8 * sizeof(int);
+// Helper to calculate modulo mask and shift bits.
+// For kNmsBoxesPerThread=32 ModuloMask will be 31, i.e 0x1F thus
+// i % 32 == i & 31. Similarly ShiftBits will be 5 so that
+// i / 32 == i >> 5. Using these bit operations should reduce the stall on host
+// thread.
+constexpr int NumBits(int n) { return (n == 0) ? 0 : NumBits(n >> 1) + 1; }
+constexpr int kNmsBoxesPerThreadModuloMask = kNmsBoxesPerThread - 1;
+constexpr int kNmsBoxesPerThreadShiftBits =
+    NumBits(kNmsBoxesPerThreadModuloMask);
+
+constexpr int kNmsBlockDim = 16;
+constexpr int kNmsBlockDimMax = 128;
+constexpr int kNmsChunkSize = 2000;
+
+template <typename T>
+__device__ EIGEN_STRONG_INLINE void Swap(T& a, T& b) {
+  T c(a);
+  a = b;
+  b = c;
+}
+
+// Check whether two boxes have an IoU greater than threshold.
+template <typename T>
+__device__ EIGEN_STRONG_INLINE bool OverThreshold(const Box* a, const Box* b,
+                                                  float a_area,
+                                                  T iou_threshold) {
+  const float b_area = (b->x2 - b->x1) * (b->y2 - b->y1);
+  if (a_area == 0.0f || b_area == 0.0f) return false;
+  const float xx1 = fmaxf(a->x1, b->x1);
+  const float yy1 = fmaxf(a->y1, b->y1);
+  const float xx2 = fminf(a->x2, b->x2);
+  const float yy2 = fminf(a->y2, b->y2);
+
+  // fdimf computes the positive difference between xx2+1 and xx1.
+  const float w = fdimf(xx2 + 1.0f, xx1);
+  const float h = fdimf(yy2 + 1.0f, yy1);
+  const float intersection = w * h;
+
+  // Testing for aa/bb > t
+  // eq with aa > bb*t (b is !=0)
+  // avoiding divisions.
+  const float aa = intersection;
+  const float bb = a_area + b_area - intersection;
+  const float bt = bb * iou_threshold;
+  return aa > bt;
+}
+
+template <bool flip_box>
+inline void Flipped(Box& box);
+template <>
+__device__ EIGEN_STRONG_INLINE void Flipped<false>(Box& box){};
+template <>
+__device__ EIGEN_STRONG_INLINE void Flipped<true>(Box& box) {
+  if (box.x1 > box.x2) Swap(box.x1, box.x2);
+  if (box.y1 > box.y2) Swap(box.y1, box.y2);
+};
+
+// If flip_box is true boxes may have x1>x2 and or y1>y2. If so change the
+// coordinates such that for all boxes x1<x2 and y1<y2. Else boxes should have
+// x1<x2 and y1<y2.
+
+// Starting from highes scoring box, mark any box which has IoU>threshold with
```

Comment:
Existing comment is about implementation details. Please add a higher level description like "Compute a bit mask `d_delete_mask` where bit `i*bit_mask_len*32+j` is set iff box i and box j has IoU score larger than specific threshold."